ee/session_recordings/session_summary/summarize_session.py

```diff
@@ -0,0 +1,147 @@
+from posthog.models import User, Team
+from posthog.session_recordings.models.session_recording import SessionRecording
+
+from posthog.session_recordings.queries.session_replay_events import SessionReplayEvents
+
+from posthog.utils import get_instance_region
+
+from ee.session_recordings.ai.utils import (
+    SessionSummaryPromptData,
+    reduce_elements_chain,
+    simplify_window_id,
+    deduplicate_urls,
+    format_dates,
+    collapse_sequence_of_events,
+)
+
+TOKENS_IN_PROMPT_HISTOGRAM = Histogram(
+    "posthog_session_summary_tokens_in_prompt_histogram",
+    "histogram of the number of tokens in the prompt used to generate a session summary",
+    buckets=[
+        0,
+        10,
+        50,
+        100,
+        500,
+        1000,
+        2000,
+        3000,
+        4000,
+        5000,
+        6000,
+        7000,
+        8000,
+        10000,
+        20000,
+        30000,
+        40000,
+        50000,
+        100000,
+        128000,
+        float("inf"),
+    ],
+)
+
+
+def summarize_recording(recording: SessionRecording, user: User, team: Team):
+    timer = ServerTimingsGathered()
+
+    with timer("get_metadata"):
+        session_metadata = SessionReplayEvents().get_metadata(session_id=str(recording.session_id), team=team)
+        if not session_metadata:
+            raise ValueError(f"no session metadata found for session_id {recording.session_id}")
+
+    with timer("get_events"):
+        session_events = SessionReplayEvents().get_events(
+            session_id=str(recording.session_id),
+            team=team,
+            metadata=session_metadata,
+            events_to_ignore=[
+                "$feature_flag_called",
+            ],
+        )
+        if not session_events or not session_events[0] or not session_events[1]:
+            raise ValueError(f"no events found for session_id {recording.session_id}")
+
+    # convert session_metadata to a Dict from a TypedDict
+    # so that we can amend its values freely
+    session_metadata_dict = dict(session_metadata)
+
+    del session_metadata_dict["distinct_id"]
+    start_time = session_metadata["start_time"]
+    session_metadata_dict["start_time"] = start_time.isoformat()
+    session_metadata_dict["end_time"] = session_metadata["end_time"].isoformat()
+
+    with timer("generate_prompt"):
+        prompt_data = deduplicate_urls(
+            collapse_sequence_of_events(
+                format_dates(
+                    reduce_elements_chain(
+                        simplify_window_id(
+                            SessionSummaryPromptData(columns=session_events[0], results=session_events[1])
+                        )
+                    ),
+                    start=start_time,
+                )
+            )
+        )
+
+    instance_region = get_instance_region() or "HOBBY"
+
+    with timer("openai_completion"):
+        result = openai.chat.completions.create(
+            # model="gpt-4-1106-preview",  # allows 128k tokens
```

Comment:
assuming this comment is for the future so you don't forget? Maybe put the model in a feature flag so you can A/B test?